dev/io.openliberty.concurrent.internal.cdi/src/io/openliberty/concurrent/internal/cdi/ConcurrencyExtension.java

```diff
@@ -1,34 +1,93 @@
+import jakarta.enterprise.concurrent.Asynchronous;
+import jakarta.enterprise.concurrent.ManagedExecutorService;
 import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.inject.Default;
+import jakarta.enterprise.inject.spi.AfterBeanDiscovery;
 import jakarta.enterprise.inject.spi.AnnotatedType;
+import jakarta.enterprise.inject.spi.Bean;
 import jakarta.enterprise.inject.spi.BeanManager;
 import jakarta.enterprise.inject.spi.BeforeBeanDiscovery;
+import jakarta.enterprise.inject.spi.CDI;
 import jakarta.enterprise.inject.spi.Extension;
+import jakarta.enterprise.inject.spi.InjectionPoint;
+import jakarta.enterprise.inject.spi.ProcessInjectionPoint;
 
-import com.ibm.ws.cdi.CDIServiceUtils;
+public class ConcurrencyExtension implements Extension {
+    private static final Set<Annotation> DEFAULT_QUALIFIER = Set.of(Default.Literal.INSTANCE);
 
-import io.openliberty.concurrent.internal.cdi.interceptor.AsyncInterceptor;
-import jakarta.enterprise.concurrent.Asynchronous;
+    /**
+     * Set of qualifier lists found on injection points.
+     */
+    private final Set<Set<Annotation>> executorQualifiers = new HashSet<>();
 
-public class ConcurrencyExtension implements Extension {
     public void beforeBeanDiscovery(@Observes BeforeBeanDiscovery beforeBeanDiscovery, BeanManager beanManager) {
         // register the interceptor binding and the interceptor
         AnnotatedType<Asynchronous> bindingType = beanManager.createAnnotatedType(Asynchronous.class);
         beforeBeanDiscovery.addInterceptorBinding(bindingType);
         AnnotatedType<AsyncInterceptor> interceptorType = beanManager.createAnnotatedType(AsyncInterceptor.class);
         beforeBeanDiscovery.addAnnotatedType(interceptorType, CDIServiceUtils.getAnnotatedTypeIdentifier(interceptorType, this.getClass()));
     }
+
+    /**
+     * Invoked for each matching injection point:
+     *
+     * @Inject @Qualifier1 @Qualifier2 ...
+     *         ManagedExecutorService executor;
+     *
+     * @param <T>         bean class that has the injection point
+     * @param event       event
+     * @param beanManager bean manager
+     */
+    public <T> void processExecutorInjectionPoint(@Observes ProcessInjectionPoint<T, ManagedExecutorService> event, BeanManager beanManager) {
+        if (ConcurrencyExtensionMetadata.eeVersion.getMajor() >= 11) {
+            // TODO check if producer already exists for the injection point and skip
+            InjectionPoint injectionPoint = event.getInjectionPoint();
+            Set<Annotation> qualifiers = injectionPoint.getQualifiers();
+            executorQualifiers.add(qualifiers);
+        }
+    }
+
+    public void afterBeanDiscovery(@Observes AfterBeanDiscovery event) {
+        if (ConcurrencyExtensionMetadata.eeVersion.getMajor() >= 11) {
+            for (Iterator<Set<Annotation>> it = executorQualifiers.iterator(); it.hasNext();) {
+                Set<Annotation> qualifiers = it.next();
+                it.remove();
```

Comment:
I'm curious as to why you remove the element here?
Doesn't seem necessary if the iterator only exists in the scope of the for loop.
But if it is I'd like to learn why :)

Reply:
Removing from the iterator also removes from the Set that is iterated over, so it's no longer holding references to it in memory.  It might have been simpler to just do a .clear operation at the end.